✨ Add Server Side Apply helper to the topology controller (#6495)
kubernetes-sigs/cluster-api: docs/book/src/developer/providers/v1.1-to-v1.2.md

  `annotations.HasPaused` and `annotations.HasSkipRemediation` respectively instead.
* `ObjectMeta.ClusterName` has been removed from `k8s.io/apimachinery/pkg/apis/meta/v1`.

### golang API Changes

- `util.ClusterToInfrastructureMapFuncWithExternallyManagedCheck` was removed and the externally managed check was added to `util.ClusterToInfrastructureMapFunc`, which required changing its signature.
   Users of the former simply need to start using the latter and users of the latter need to add the new arguments to their call.

### Required API Changes for providers

- ClusterClass and managed topologies are now using [Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) 

>   to proper manager other controllers like CAPA/CAPZ coauthoring slices, see [#6320](https://github.com/kubernetes-sigs/cluster-api/issues/6320).

Reviewer: ```suggestion
  to properly manage other controllers like CAPA/CAPZ coauthoring slices, see [#6320](https://github.com/kubernetes-sigs/cluster-api/issues/6320).
```

>   In order to proper leverage this feature providers are required to add marker to their API types as described in

Reviewer: ```suggestion
  In order to properly leverage this feature providers are required to add markers to their API types as described in
```

  [merge-strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
  NOTE: the change will cause a rollout on existing clusters created with ClusterClass

E.g. in CAPA

```go
// +optional
Subnets Subnets `json:"subnets,omitempty"
```
Must be modified into:

```go
// +optional

> // +listType=Map

Reviewer: ```suggestion
// +listType=map
```
afaik
